### ISSUE TITLE
Combining dieresis not seen as superior to dieresis, even if “Glyph Info→Components” used

### ISSUE THREAD

**Issue by ctrlcctrlv** [fontforge/fontforge#3708]
A combining dieresis should always be seen as superior to a regular dieresis for the purpose of glyph building.

Even if you don't agree, certainly it should be seen as superior when custom components are in use!

Yet it isn't, see [DieresisIssue.zip](https://github.com/fontforge/fontforge/files/3219364/DieresisIssue.zip) for a minimal example.

**Comment by skef** [comment]
Blarg. 

Reading through the code, this appears to result from the use of the `accents` array in `GetGoodAccentGlyph()`, which is called from `SCCenterAccent()`. GW's explanation (from 2001-6) is the following:
```
/* for accents between 0x300 and 345 these are some synonyms */
/* type1 wants accented chars built with accents in the 0x2c? range */
/*  except for grave and acute which live in iso8859-1 range */
/*  this table is ordered on a best try basis */
```
The best contemporary argument for continuing this practice would be: "PostScript Type 1 fonts won't work unless constructed in this way. Other fonts will work regardless of how they're constructed (perhaps, in some cases, after certain references are unlinked). Therefore it is preferable to leave this heuristic in place."

At this late date one might make this alternative argument: "Type 1 fonts are now used only rarely. It should be possible to create and edit such fonts with FontForge, but perhaps with certain restrictions and complications. One of which might be requiring that the combining marks be unmapped when creating precombined characters."

That would sanction reordering of the `accents` entries to put the combining characters first. 

The other problem is how late that array is applied. Assuming it's best to avoid a significant reorganization of code, one could add a boolean parameter to the relevant calls to control whether it is used, and skip the substitution when @ctrlcctrlv's explicit override is present.

**Comment by ctrlcctrlv** [comment]
@skef What if we just unlink the references to a combining dieresis or combining grave when writing a Type1 font? I could probably whip up a PR to both fix this and do just that if you accept this as a valid solution. After all, there is precedent for it; when saving a TTF/OTF if a reference is transformed it's automatically unlinked.

**Comment by frank-trampe** [comment]
@ctrlcctrlv, Type 1 fonts support those combining marks, no?

**Comment by frank-trampe** [comment]
@skef, do you have any thoughts on how best to expose the over-ride?

**Comment by ctrlcctrlv** [comment]
If they do support it correctly this override wouldn't be there in the first place.

**Comment by ctrlcctrlv** [comment]
Unsupported references in the target format should just be unlinked. That's what we do with TTF... No reason not to do it with some obscure format I've never even used. Maybe wasn't obscure when FontForge was in development, but then again neither was ActiveX or OS/2 at that time.

**Comment by frank-trampe** [comment]
The unsupported references would only be there as a result of the other change, though, right?

**Comment by frank-trampe** [comment]
What if we change the priority but then have a conditional substitution for Type 1 output? And was there ever a day when OS/2 was not obscure?

**Comment by ctrlcctrlv** [comment]
The other change is necessary because it is completely unintuitive for users for combining tilde to be superior to tilde but combining dieresis isn't superior to dieresis.

**Comment by frank-trampe** [comment]
Yes, but it has the effect of breaking something concrete that worked before, so the combined changes would be regressive.

**Comment by frank-trampe** [comment]
I need to get to sleep, but I'll be back tomorrow.

**Comment by ctrlcctrlv** [comment]
No they aren't. There's absolutely no difference for users of a type 1 font if it has references or not. Type 1 is not SFD, it's not a first class format that should be edited.

**Comment by ctrlcctrlv** [comment]
🙄

**Comment by skef** [comment]
> @skef, do you have any thoughts on how best to expose the over-ride?

> What if we change the priority but then have a conditional substitution for Type 1 output? And was there ever a day when OS/2 was not obscure?

So there are maybe two ideas here, one being a special "mode" for *constructing* Type 1 fonts, and the other being some reference alterations when *outputting* Type 1 fonts. The first idea has the advantage that we don't need to understand the relevant Type 1 limitations -- FontForge could just offer the current heuristic as an option. The second has the advantages of (maybe?) not requiring any additional UI and not raising the problem that basically no one will realize they needed to set our added parameter until it's "too late". 

There is an important disadvantage to idea 1, however. Consider these two use cases:

1. A designer wants to create a new Type 1 font.
2. A user wants to convert an existing font (or a subset of it) to Type 1 format to use with older software or hardware. 

I suspect that, at this point in the evolution of font formats, the second case is much more likely than the first case, and FontForge is an obvious tool, maybe *the* obvious tool, to grab for that purpose. So it would be better going forward to alter the code to handle the problem at conversion time. 

If that's right, the technical question this issue raises is *what are the actual Type 1 limitations?* If I'm reading it right [this blog entry](https://blog.typekit.com/2005/10/06/phasing_out_typ/) suggests that a Type 1 font can contain at most 256 mapped glyphs and some number of additional glyphs. GW's code doesn't *enforce* the preference -- if the sub-256 glyphs are empty he uses the combining characters. That suggests the relevant issue is either a hard limit of 256 or a desire to avoid adding unmapped glyphs when possible, probably for RAM reasons. Going on this information alone, two strategies suggest themselves:

1. *If* it's OK to have some additional unmapped glyphs, one could simply add the additional combining characters to the font when needed, so that the references can be pulled from them. (Assuming it's possible, in the format, to pull references from an unmapped glyph.) This may cause problems for a few people using very old equipment (for RAM reasons) but for everyone else a little extra memory won't be a problem. Someone constructing a font for ancient uses will have some extra work to do. 
2. If it's *not* OK, or if we just want to be conservative, the output code will have to attempt actual substitutions in those cases when the sub-256 characters are present in the font. My recommendation would be to verify that the contours are identical and put up a dialog with a cancel option if they are not. (The code to check whether contours are equivalent (rather than identical) is currently flaky.) 

That's a lot of bloviation on my part. Does FontForge have a Type 1 expert to consult, or could it dig one up?

**Comment by ctrlcctrlv** [comment]
Dissolving the references makes even more sense in light of what you wrote, given the 256 character limit&mdash;that's what #3712 does. I needed this fix for a font I'm working on and have no interest in writing up any other solution. If someone is making a brand new type 1 font they shouldn't add the combining characters anyway, as space is precious given the hard limit of 256 characters they have.

**Comment by ctrlcctrlv** [comment]
Closing as resolved by #3817.
